### DISCUSSION TITLE
Can someone help with compiling `PcapPlusPlus`?

### DISCUSSION THREAD

**Discussion by IsaacsLab42** [cpm-cmake/CPM.cmake#565] [Q&A]
I'm trying to use CPM to compile and statically link both `libpcap` and `PcapPlusPlus`, but I'm having problems. My `CMakeLists.txt` file looks like this:

```cmake
CPMAddPackage(
    NAME libpcap
    GITHUB_REPOSITORY the-tcpdump-group/libpcap
    GIT_TAG libpcap-1.10.4
    OPTIONS
        "BUILD_SHARED_LIBS OFF"
        "BUILD_WITH_LIBNL OFf"
        "DISABLE_BLUETOOTH ON"
        "DISABLE_DAG OFF"
        "DISABLE_DBUS ON"
        "DISABLE_LINUX_USBMON ON"
        "DISABLE_NETMAP ON"
        "DISABLE_RDMA ON"
)

CPMAddPackage(
    NAME pcapplusplus
    GITHUB_REPOSITORY seladb/PcapPlusPlus
    GIT_TAG v23.09
    OPTIONS
        "BUILD_SHARED_LIBS OFF"
        "PCAPPP_USE_PF_RING OFF"
        "PCAPPP_USE_DPDK OFF"
        "PCAPPP_ENABLE_PCAP_IMMEDIATE_MODE ON"
        "PCAPPP_ENABLE_PCAP_SET_DIRECTION ON"
)
```

The error I get while PcapPlusPlus is configuring is:

```
PCAP library not found
```

So, I tried adding this to the `CMakeLists.txt` file in-between the two `CPMAddPackage` blocks:

```
if(libpcap_ADDED)
    set(PCAP_ROOT "${libpcap_BINARY_DIR}" CACHE STRING "" FORCE)
    set(PCAP_INCLUDE_DIR "${libpcap_SOURCE_DIR}" CACHE STRING "" FORCE)
    set(PCAP_LIBRARY "${libpcap_BINARY_DIR}/libpcap.a" CACHE STRING "" FORCE)
    # I've also tried set(PCAP_LIBRARY pcap_static CACHE STRING "" FORCE)
endif()
```

Now, the error changes to:

```
Couldn't determine how to link against libpcap
```

I finally managed to track down the cause of the error, as logged in the `CMakeConfigureLog.yaml` file:

```
ninja: error: '/home/user/src/build/debug/_deps/libpcap-build/libpcap.a', needed by 'cmTC_669c0', missing and no known rule to make it
```

Makes sense, since this is still the configure stage and `libpcap.a` has not been built yet. Is there any way to get around this problem? I'm totally onboard with adding a `PATCH_COMMAND` to munge one of the CMake files in either project, if that's what it takes. It just seems like there should be some other non-invasive solution though.

**Comment by k-channon-PA** [comment]
Have you tried running your cmake command with the `--debug-output` flag? Or even the `--trace` flag, if you want to see ALL THE THINGS!

You might be able to understand better what's causing the error with that output.

**Reply by IsaacsLab42** [reply]
Yes, I tried various traces and debug outputs. As I said, the problem is that the build is looking for libpcap.a to be there, before it has been built.
